src/behavior_generation_lecture_python/graph_search/a_star.py

```diff
@@ -1,160 +1,172 @@
 
-        def retrace_path(node):
-            path = [node]
+        open_set.add(start)
+        self.nodes_dict[start].cost_to_come = 0
 
-            while node.predecessor is not None:
-                node = node.predecessor
-                path.append(node)
+        while open_set:
+            current_node = extract_min(node_set=open_set, node_dict=self.nodes_dict)
 
-            path.reverse()
-            return path
+            if current_node == end:
+                return True
 
-        while open_set:
-            x = extract_min(open_set)
-            closed_set.append(x)
-
-            if x.node == end:
-                return retrace_path(x)
-            else:
-                for node in x.node.connected_to:
-                    x_tilde = AStarNode(node, end)
-
-                    if x_tilde.node not in closed_set:
-                        cost = x.C + x_tilde.node.distance_to(x.node)
-
-                        if (
-                            not x_tilde.node in [x.node for x in open_set]
-                            or cost < x_tilde.G
-                        ):
-                            x_tilde.predecessor = x
-                            x_tilde.C = cost
-                            x_tilde.J = x_tilde.C + x_tilde.G
-
-                            if not x_tilde.node in [x.node for x in open_set]:
-                                open_set.append(x_tilde)
-
-    def draw_result(self, result):
-        for i in range(len(result) - 1):
-            node_from = result[i].node
-            node_to = result[i + 1].node
-            self.ax.plot(
-                [node_from.x, node_to.x],
-                [node_from.y, node_to.y],
-                linewidth=2,
-                color="b",
-            )
+            closed_set.add(current_node)
 
-            distance = node_from.distance_to(node_to)
-            x_mid = (node_from.x + node_to.x) / 2
-            y_mid = (node_from.y + node_to.y) / 2
-            self.ax.annotate(f"{distance:.2f}", (x_mid + 10, y_mid + 10))
+            for successor_node in self.nodes_dict[current_node].connected_to:
+                if successor_node in closed_set:
+                    continue
 
-        plt.show()
+                tentative_cost_to_come = self.nodes_dict[
+                    current_node
+                ].cost_to_come + np.linalg.norm(
+                    self.nodes_dict[current_node].position
+                    - self.nodes_dict[successor_node].position
+                )
+                if (
+                    successor_node in open_set
+                    and tentative_cost_to_come
+                    >= self.nodes_dict[successor_node].cost_to_come
+                ):
+                    continue
+
+                self.nodes_dict[successor_node].predecessor = current_node
+                self.nodes_dict[successor_node].cost_to_come = tentative_cost_to_come
+                open_set.add(successor_node)
+
+        return False
```

Comment:
How could this return `False`?

Reply:
If the graph is not connected/ if there is no path from start to end? Does that make sense?

Reply:
Yes that makes sense. And maybe that is even a common case in practical applications. Especially if the graph and start/end nodes are not chosen by a user, but another function.